src/directives/repeat.ts

```diff
@@ -41,67 +96,139 @@ export function repeat<T>(
-        }
+    // New result list is eagerly generated from items and marked with its key.
+    // This is saved in partMapCache at the end of the update.
+    const newResults: Array<KeyedTemplateResult> = [];
+    let index = 0;
+    for (let item of items) {
+      const key = keyFn ? keyFn(item) : index;
+      if (newKeyToIndexMap.has(key)) {
+        console.warn(`Duplicate key '${
+            key}' detected in repeat; behavior is undefined.`);
       } else {
-        // else part is in the correct position already
-        currentMarker = itemPart.endNode.nextSibling!;
+        const result = newResults[index] = template !(item, index);
+        newKeyToIndexMap.set(key, index);
+        result.key = key;
+        index++;
       }
-
-      itemPart.setValue(result);
-      itemPart.commit();
     }
 
-    // Cleanup
-    if (currentMarker !== part.endNode) {
-      removeNodes(container, currentMarker, part.endNode);
-      keyMap.forEach(cleanMap);
+    // Head and tail pointers to old parts and new results
+    let oldStartIndex = 0;
+    let oldStartPart = oldParts[0];
+    let oldEndIndex = oldParts.length - 1;
+    let oldEndPart = oldParts[oldEndIndex];
+    let newStartIndex = 0;
+    let newStartResult = newResults[0];
+    let newEndIndex = newResults.length - 1;
+    let newEndResult = newResults[newEndIndex];
+
+    // Overview of O(n) reconciliation algorithm (general approach based on
+    // ideas found in ivi, vue, snabdom, etc.):
+    // * Iterate old & new lists from both sides, updating, swapping, or
+    // removing
+    //   items at the head/tail locations until neither head nor tail can move
+    // * Once head and tail cannot move, any mismatches are due to either new or
+    //   moved items; create and insert new items, or else move old part to new
+    //   position using the "old key to old index" map.
+    // * If either the new or old pointers move past each other then all we have
+    //   left is additions (if old list exhsusted) or removals (if new list
+    //   exhausted). Those are handled in the final while loops at the end.
+    // * TODO(kschaaf) Note, we could calculate the longest increasing
+    //   subsequence (LIS) of old items in new position, and only move those not
+    //   in the LIS set. However that costs O(nlogn) time and adds a bit more
+    //   code, and only helps make rare types of mutations require fewer moves.
+    //   The above handles removes, adds, reversal, swaps, and single moves of
+    //   contiguous items in linear time, in the minimum number of moves. As
+    //   the number of multiple moves where LIS might help approaches a random
+    //   shuffle, the LIS optimization becomes less helpful, so it seems not
+    //   worth the code at this point, but could consider if a compelling case
+    //   arises.
+    while (oldStartIndex <= oldEndIndex && newStartIndex <= newEndIndex) {
+      if (oldStartPart == null) {
```

Comment:
Right, spots in the old array are nulled out as they are moved in the last else clause.  Steve had a similar suggestion to add a comment that that's what's going on, since the moves come at the end and a new reader sees them out of order.